cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo

```diff
@@ -151,152 +286,1003 @@ fn calculate_iteration_count(exp_length: u64, exp_highp: u256) -> u64 {
+    #[test]
+    fn test_modexp_precompile_input_output_worst() {
+        let mut calldata = array![];
+        let l0f: u128 = Into::<_, felt252>::into(PREV_PRIME_384_M2.limb0).try_into().unwrap();
+        let l1f: u128 = Into::<_, felt252>::into(PREV_PRIME_384_M2.limb1).try_into().unwrap();
+        let size = array![48_u8].span().pad_left_with_zeroes(32);
+        calldata.append_span(size);
+        calldata.append_span(size);
+        calldata.append_span(size);
+
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l0f.to_be_bytes());
+
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span((l0f + 1).to_be_bytes());
+
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span(l1f.to_be_bytes());
+        calldata.append_span((l0f + 2).to_be_bytes());
+
+        let (gas, result) = ModExp::exec(calldata.span()).unwrap();
+        let expected_result = array![1].span().pad_left_with_zeroes(48);
+        let expected_gas = 4596;
+        assert_eq!(result, expected_result);
+        assert_eq!(gas, expected_gas);
+    }
+
+    #[test]
+    fn test_modexp_precompile_size_gt_48_should_err() {
+        let mut calldata = array![];
+        let size = array![49_u8].span().pad_left_with_zeroes(32);
+        calldata.append_span(size);
+        calldata.append_span(size);
+        calldata.append_span(size);
+
+        let result = ModExp::exec(calldata.span());
+        assert!(result.is_err());
+    }
 
-        let expected_gas = 44_954;
+    #[test]
+    fn test_modexp_precompile_base_and_exp_zero_should_return_empty() {
+        let mut calldata = array![];
+        let size_zero = array![0].span().pad_left_with_zeroes(32);
+        let size_not_zero = array![48].span().pad_left_with_zeroes(32);
 
-        let gas_before = vm.gas_left;
-        Precompiles::exec_precompile(ref vm).unwrap();
-        let gas_after = vm.gas_left;
+        calldata.append_span(size_zero);
+        calldata.append_span(size_not_zero);
+        calldata.append_span(size_zero);
 
-        assert_eq!(gas_before - gas_after, expected_gas);
-        assert_eq!(vm.return_data, expected);
+        let (gas, result) = ModExp::exec(calldata.span()).unwrap();
+        let expected_result = [].span();
+        let expected_gas = 200;
+        assert_eq!(result, expected_result);
+        assert_eq!(gas, expected_gas);
     }
 
+    // To test all input sizes, we use the fact that:
+    // For prime p, a^(p-1) mod p = 1, for all a. (Fermat's little theorem)
+    // Using prime_deltas we get a prime_i = 256^(i+2) + prime_deltas[i] of size i+2 bytes.
+    // We can then use p = prime_i and a = p - 2 and expect the result to be 1 with padding
```

Comment:
What are prime deltas ?
This explaination is not clear to me could you explain in more details ?

Reply:
from what I understand:
- 256^(i+2) gives a number that's exactly i+2 bytes
- prime_deltas[i] is a small "adjustment" value to adjust the number to a prime number

it's actually a super clever test strategy 🫨 better than bruteforce with hypotheses 😓.